File: iOS/Advanced_Demo/Advanced/Feature/Room/View/BeautyBarView.m
//
//  BeautyBarView.m
//  quickstart

#import "BeautyBarView.h"
#import "EffectButton.h"
#import "Masonry.h"
#import "BeautyModel.h"

@interface BeautyBarView ()
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UISegmentedControl *segment;

@property (nonatomic, strong) UILabel *sliderLabel;
@property (nonatomic, strong) UISlider *slider;

@property (nonatomic, strong) UIView *buttonListView;

@property (nonatomic, strong) NSMutableArray <NSArray <EffectModel *> *> *dataArray;
@end

@implementation BeautyBarView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.backgroundColor = [UIColor whiteColor];

        [self addSubview:self.titleLabel];
        [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
          make.left.mas_equalTo(20);
          make.top.mas_equalTo(45);
        }];

        [self addSubview:self.segment];
        [self.segment mas_makeConstraints:^(MASConstraintMaker *make) {
          make.top.equalTo(self.titleLabel.mas_bottom).offset(20);
          make.left.mas_equalTo(20);
          make.right.mas_equalTo(-20);
        }];

        [self addSubview:self.sliderLabel];
        [self.sliderLabel mas_makeConstraints:^(MASConstraintMaker *make) {
          make.left.mas_equalTo(20);
          make.top.equalTo(self.segment.mas_bottom).offset(35);
        }];

        [self addSubview:self.slider];
        [self.slider mas_makeConstraints:^(MASConstraintMaker *make) {
          make.centerY.equalTo(self.sliderLabel);
          make.left.equalTo(self.sliderLabel.mas_right).offset(30);
          make.right.mas_equalTo(-20);
        }];

        [self addSubview:self.buttonListView];
        [self.buttonListView mas_makeConstraints:^(MASConstraintMaker *make) {
          make.left.mas_equalTo(20);
          make.top.equalTo(self.segment.mas_bottom).offset(90);
          make.bottom.equalTo(self);
          make.right.mas_offset(-20);
        }];

        [self loadData];
    }
    return self;
}

- (void)loadData {
    NSMutableArray *jsonList = [NSMutableArray array];
    [jsonList addObject:[BeautyModel beautyList]];
    [jsonList addObject:[BeautyModel stickerList]];
    [jsonList addObject:[BeautyModel filterList]];
    [jsonList addObject:[BeautyModel virtualBackgroundList]];
    
    for (int i = 0; i < jsonList.count; i++) {
        NSArray *sectionList = jsonList[i];
        NSMutableArray *modelList = [NSMutableArray array];
        for (int j = 0; j < sectionList.count; j++) {
            EffectModel *model = [[EffectModel alloc] init];
            model.type = i;
            model.selected = NO;
            model.value = 0;
            model.dic = sectionList[j];
            [modelList addObject:model];
        }
        [self.dataArray addObject:[modelList copy]];
    }
    [self segmentValueChanged:nil];
}

- (void)layoutSubviews {
    [super layoutSubviews];

    UIBezierPath *cornerRadiusPath = [UIBezierPath bezierPathWithRoundedRect:self.bounds byRoundingCorners:UIRectCornerTopRight | UIRectCornerTopLeft cornerRadii:CGSizeMake(32, 32)];
    CAShapeLayer *cornerRadiusLayer = [[CAShapeLayer alloc] init];

    cornerRadiusLayer.frame = self.bounds;

    cornerRadiusLayer.path = cornerRadiusPath.CGPath;

    self.layer.mask = cornerRadiusLayer;
}

- (void)buttonClicked:(UIButton *)button {
    for (EffectButton *subButton in self.buttonListView.subviews) {
        if (subButton == button) {
            if (self.segment.selectedSegmentIndex == 1) {
                //贴纸
                subButton.selected = !subButton.selected;
                subButton.model.selected = subButton.selected;
            } else if (self.segment.selectedSegmentIndex == 2) {
                //滤镜
                subButton.selected = YES;
                subButton.model.selected = YES;
                subButton.model.value = 0;
                self.slider.value = subButton.model.value;
            } else if (self.segment.selectedSegmentIndex == 0) {
                //美颜
                subButton.selected = YES;
                subButton.model.selected = YES;
                self.slider.value = subButton.model.value;
            } else if (self.segment.selectedSegmentIndex == 3) {
                //背景分割
                subButton.selected = !subButton.selected;
                subButton.model.selected = subButton.selected;
            } else {
                //error
            }

            if ([self.delegate respondsToSelector:@selector(beautyBarView:didClickedEffect:)]) {
                [self.delegate beautyBarView:self didClickedEffect:subButton.model];
            }
        } else {
            subButton.selected = NO;
            subButton.model.selected = NO;
        }
    }
}

- (void)sliderValueChanged:(UISlider *)slider {
    for (EffectButton *button in self.buttonListView.subviews) {
        if (button.selected) {
            button.model.value = slider.value;
            if ([self.delegate respondsToSelector:@selector(beautyBarView:didChangeEffectValue:)]) {
                [self.delegate beautyBarView:self didChangeEffectValue:button.model];
            }
        }
    }
}

#pragma mark - private action
- (void)segmentValueChanged:(UISegmentedControl *)segment {
    NSArray *modelArray = self.dataArray[self.segment.selectedSegmentIndex];

    for (UIView *subView in self.buttonListView.subviews) {
        [subView removeFromSuperview];
    }

    for (int i = 0; i < modelArray.count; i++) {
        EffectModel *model = modelArray[i];
        EffectButton *button = [[EffectButton alloc] init];
        button.model = model;

        [button addTarget:self action:@selector(buttonClicked:) forControlEvents:UIControlEventTouchUpInside];

        [self.buttonListView addSubview:button];

        NSInteger row = i / 4;
        NSInteger col = i % 4;
        [button mas_makeConstraints:^(MASConstraintMaker *make) {
          make.top.mas_equalTo(row * (30 + 36));
          make.centerX.equalTo(self.buttonListView).multipliedBy(0.5 * col + 0.25);
          make.width.mas_equalTo(70);
          make.height.mas_equalTo(36);
        }];
    }

    switch (segment.selectedSegmentIndex) {
        case 0: //美颜
        {
            [self.buttonListView mas_updateConstraints:^(MASConstraintMaker *make) {
              make.top.equalTo(self.segment.mas_bottom).offset(90);
            }];

            self.sliderLabel.text = @"美颜强度";
            self.sliderLabel.hidden = NO;
            self.slider.hidden = NO;

            EffectButton *button = self.buttonListView.subviews.firstObject;
            button.selected = YES;

            EffectModel *model = modelArray[0];

            self.slider.value = model.value;

        } break;
        case 1: //贴纸
        {
            [self.buttonListView mas_updateConstraints:^(MASConstraintMaker *make) {
              make.top.equalTo(self.segment.mas_bottom).offset(36);
            }];

            self.sliderLabel.hidden = YES;
            self.slider.hidden = YES;

            for (EffectButton *button in self.buttonListView.subviews) {
                button.selected = button.model.selected;
            }
        } break;
        case 2: //滤镜
        {
            [self.buttonListView mas_updateConstraints:^(MASConstraintMaker *make) {
              make.top.equalTo(self.segment.mas_bottom).offset(90);
            }];

            self.sliderLabel.text = @"滤镜强度";
            self.sliderLabel.hidden = NO;
            self.slider.hidden = NO;

            // 之前选中过
            BOOL selected = NO;
            for (EffectButton *button in self.buttonListView.subviews) {
                if (button.model.selected) {
                    button.selected = YES;
                    self.slider.value = button.model.value;
                    selected = YES;
                }
            }

            if (selected == NO) {
                //之前没有选中过,默认第一个
                EffectButton *button = self.buttonListView.subviews.firstObject;
                button.selected = YES;

                EffectModel *model = modelArray[0];
                self.slider.value = model.value;
            }
        } break;
        case 3: //背景分割
        {
            [self.buttonListView mas_updateConstraints:^(MASConstraintMaker *make) {
              make.top.equalTo(self.segment.mas_bottom).offset(36);
            }];

            self.sliderLabel.hidden = YES;
            self.slider.hidden = YES;

            // 之前选中过
            BOOL selected = NO;
            for (EffectButton *button in self.buttonListView.subviews) {
                if (button.model.selected) {
                    button.selected = YES;
                    selected = YES;
                }
            }

            if (selected == NO) {
                //之前没有选中过,默认第一个
                EffectButton *button = self.buttonListView.subviews.firstObject;
                button.selected = NO;
            }
        } break;
    }
}

#pragma mark - getter

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont systemFontOfSize:28];
        _titleLabel.text = @"特效";
    }
    return _titleLabel;
}

- (UISegmentedControl *)segment {
    if (!_segment) {
        _segment = [[UISegmentedControl alloc] initWithItems:[BeautyModel sectionList]];
        _segment.selectedSegmentIndex = 0;
        [_segment addTarget:self action:@selector(segmentValueChanged:) forControlEvents:UIControlEventValueChanged];
    }
    return _segment;
}

- (UISlider *)slider {
    if (!_slider) {
        _slider = [[UISlider alloc] init];
        [_slider addTarget:self action:@selector(sliderValueChanged:) forControlEvents:UIControlEventValueChanged];
    }
    return _slider;
}

- (UIView *)buttonListView {
    if (!_buttonListView) {
        _buttonListView = [[UIView alloc] init];
    }
    return _buttonListView;
}

- (UILabel *)sliderLabel {
    if (!_sliderLabel) {
        _sliderLabel = [[UILabel alloc] init];
        _sliderLabel.font = [UIFont systemFontOfSize:17];
        _sliderLabel.text = @"美颜强度";
    }
    return _sliderLabel;
}

- (NSMutableArray<NSArray <EffectModel *> *> *)dataArray {
    if (!_dataArray) {
        _dataArray = [[NSMutableArray alloc] init];
    }
    return _dataArray;
}
@end

File: iOS/ApiExample/ApiExample/ImportantComponents/Beauty/VolcBeauty/BeautyBarView.swift
//
//  BeautyBarView.swift
//  ApiExample
//
//  Created by bytedance on 2023/10/31.
//  Copyright © 2021 bytedance. All rights reserved.
//

import UIKit
import SnapKit

protocol BeautyBarViewDelegate: NSObjectProtocol {
    func beautyBarView(barView: BeautyBarView, didClickedEffect model: EffectModel)
    func beautyBarView(barView: BeautyBarView, didChangeEffectValue model: EffectModel)
}

class BeautyBarView: UIView {
    
    weak var delegate: BeautyBarViewDelegate?
    
    var segment: UISegmentedControl!
    var sliderLabel: UILabel!
    var slider: UISlider!
    var buttonListView: UIView!
    
    var dataArray = [[EffectModel]]()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.white
        
        // segment
        segment = UISegmentedControl(items: BeautyModel.sectionList())
        segment.selectedSegmentIndex = 0
        segment.addTarget(self, action: #selector(segmentValueChanged), for: .valueChanged)
        self.addSubview(segment)
        segment.snp.makeConstraints { make in
            make.top.equalToSuperview()
            make.left.equalTo(20)
            make.right.equalTo(-20)
        }
        
        // sliderLabel
        sliderLabel = UILabel()
        sliderLabel.font = UIFont.systemFont(ofSize: 17)
        sliderLabel.text = "美颜强度"
        self.addSubview(sliderLabel)
        sliderLabel.snp.makeConstraints { make in
            make.left.equalTo(20)
            make.top.equalTo(segment.snp.bottom).offset(35)
        }
        
        // slider
        slider = UISlider()
        slider.addTarget(self, action: #selector(sliderValueChanged), for: .valueChanged)
        self.addSubview(slider)
        slider.snp.makeConstraints { make in
            make.centerY.equalTo(sliderLabel)
            make.left.equalTo(sliderLabel.snp.right).offset(30)
            make.right.equalTo(-20)
        }
        
        // buttonListView
        buttonListView = UIView()
        self.addSubview(buttonListView)
        buttonListView.snp.makeConstraints { make in
            make.left.equalTo(20)
            make.top.equalTo(segment.snp.bottom).offset(90)
            make.bottom.equalTo(self)
            make.right.equalTo(-20)
        }
        
        loadData()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    // MARK: - Private action
    @objc func buttonClicked(button: UIButton) {
        for case let subButton as EffectButton in buttonListView.subviews {
            if subButton == button {
                switch segment.selectedSegmentIndex {
                case 0:  //美颜
                    subButton.isSelected = true
                    subButton.model!.selected = true
                    slider.value = subButton.model!.value
                case 1:  //滤镜
                    subButton.isSelected = true
                    subButton.model!.selected = true
                    subButton.model!.value = 0
                    slider.value = subButton.model!.value
                case 2, 3:  //贴纸, 背景分割
                    subButton.isSelected = !subButton.isSelected
                    subButton.model!.selected = subButton.isSelected
                default:
                    // error
                    break
                }
                
                (delegate)?.beautyBarView(barView: self, didClickedEffect: subButton.model!)
            } else {
                subButton.isSelected = false
                subButton.model!.selected = false
            }
        }
    }
    
    
    @objc func segmentValueChanged(segment: UISegmentedControl) {
        let modelArray = dataArray[segment.selectedSegmentIndex]
        
        buttonListView.subviews.forEach { $0.removeFromSuperview() }
        
        for i in 0..<modelArray.count {
            let model = modelArray[i]
            let button = EffectButton()
            button.model = model
            
            button.addTarget(self, action: #selector(buttonClicked), for: .touchUpInside)
            
            buttonListView.addSubview(button)
            
            let row = i / 4
            let col = i % 4
            button.snp.makeConstraints { make in
                make.top.equalTo(row * (30 + 36))
                make.centerX.equalTo(buttonListView).multipliedBy(0.5 * CGFloat(col) + 0.25)
                make.width.equalTo(70)
                make.height.equalTo(36)
            }
        }
        
        switch segment.selectedSegmentIndex {
        case 0:  //美颜
            buttonListView.snp.updateConstraints { make in
                make.top.equalTo(segment.snp.bottom).offset(90)
            }
            sliderLabel.text = "美颜强度"
            sliderLabel.isHidden = false
            slider.isHidden = false
            
            if let button = buttonListView.subviews.first as? EffectButton {
                button.isSelected = true
                let model = modelArray[0]
                slider.value = Float(model.value)
            }
            
        case 1:  //滤镜
            buttonListView.snp.updateConstraints { make in
                make.top.equalTo(segment.snp.bottom).offset(90)
            }
            sliderLabel.text = "滤镜强度"
            sliderLabel.isHidden = false
            slider.isHidden = false
            
            var selected = false
            for case let button as EffectButton in buttonListView.subviews {
                if button.model!.selected {
                    button.isSelected = true
                    slider.value = Float(button.model!.value)
                    selected = true
                }
            }
            
            if !selected {
                //之前没有选中过,默认第一个
                if let button = buttonListView.subviews.first as? EffectButton {
                    button.isSelected = true
                    let model = modelArray[0]
                    slider.value = Float(model.value)
                }
            }
            
        case 2:  //贴纸
            buttonListView.snp.updateConstraints { make in
                make.top.equalTo(segment.snp.bottom).offset(36)
            }
            sliderLabel.isHidden = true
            slider.isHidden = true
            
            for case let button as EffectButton in buttonListView.subviews {
                button.isSelected = button.model!.selected
            }
            
        case 3:  //背景分割
            buttonListView.snp.updateConstraints { make in
                make.top.equalTo(segment.snp.bottom).offset(36)
            }
            sliderLabel.isHidden = true
            slider.isHidden = true
            
            var selected = false
            for case let button as EffectButton in buttonListView.subviews {
                if button.model!.selected {
                    button.isSelected = true
                    selected = true
                }
            }
            
            if !selected {
                //之前没有选中过,默认第一个
                if let button = buttonListView.subviews.first as? EffectButton {
                    button.isSelected = false
                }
            }
            
        default:
            break
        }
    }
    
    @objc func sliderValueChanged(slider: UISlider) {
        for case let button as EffectButton in buttonListView.subviews {
            if button.isSelected {
                button.model!.value = slider.value
                (delegate)?.beautyBarView(barView: self, didChangeEffectValue: button.model!)
            }
        }
    }
    
    func loadData() {
        var jsonList = [[Dictionary<String, Any>]]()
        jsonList.append(BeautyModel.beautyList())
        jsonList.append(BeautyModel.filterList())
        jsonList.append(BeautyModel.stickerList())
        jsonList.append(BeautyModel.virtualBackgroundList())
        
        for i in 0..<jsonList.count {
            let sectionList = jsonList[i]
            var modelList = [EffectModel]()
            for j in 0..<sectionList.count {
                let model = EffectModel()
                model.type = EffectModelType(rawValue: i)!
                model.selected = false
                model.value = 0
                model.name = sectionList[j]["name"] as? String
                model.key = sectionList[j]["key"] as? String
                model.subType = sectionList[j]["subType"] as? EffectModelSubType
                modelList.append(model)
            }
            dataArray.append(modelList)
        }
        
        segmentValueChanged(segment: segment)
    }
    
}
